cpp/src/rolling/range_window_bounds_detail.hpp

```diff
@@ -0,0 +1,139 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/rolling/range_window_bounds.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/types.hpp>
+#include <cudf/wrappers/durations.hpp>
+
+namespace cudf {
+namespace detail {
+
+/// Checks if the specified type is supported in a range_window_bounds.
+template <typename RangeType>
+constexpr bool is_supported_range_type()
+{
+  return cudf::is_duration<RangeType>() ||
+         (std::is_integral<RangeType>::value && !cudf::is_boolean<RangeType>());
+}
+
+/// Checks if the specified type is a supported target type,
+/// as an orderby column, for comparisons with a range_window_bounds scalar.
+template <typename ColumnType>
+constexpr bool is_supported_order_by_column_type()
+{
+  return cudf::is_timestamp<ColumnType>() ||
+         (std::is_integral<ColumnType>::value && !cudf::is_boolean<ColumnType>());
+  ;
+}
+
+/// Range-comparable representation type for an orderby column type.
+/// This is the datatype used for range comparisons.
+///   1. For integral orderby column types `T`, comparisons are done as `T`.
+///      E.g. `range_type_for<int32_t>` == `int32_t`.
+///   2. For timestamp orderby columns:
+///      a. For `TIMESTAMP_DAYS`, the range-type is `DURATION_DAYS`.
+///         Comparisons are done in `int32_t`.
+///      b. For all other timestamp types, comparisons are done in `int64_t`.
+template <typename ColumnType, typename = void>
+struct range_type_impl {
+  using type     = void;
+  using rep_type = void;
+};
+
+template <typename ColumnType>
+struct range_type_impl<
+  ColumnType,
+  std::enable_if_t<std::is_integral<ColumnType>::value && !cudf::is_boolean<ColumnType>(), void>> {
+  using type     = ColumnType;
+  using rep_type = ColumnType;
+};
+
+template <typename TimestampType>
+struct range_type_impl<TimestampType, std::enable_if_t<cudf::is_timestamp<TimestampType>(), void>> {
+  using type     = typename TimestampType::duration;
+  using rep_type = typename type::rep;
+};
+
+template <typename ColumnType>
+using range_type = typename range_type_impl<ColumnType>::type;
+
+template <typename ColumnType>
+using range_rep_type = typename range_type_impl<ColumnType>::rep_type;
+
+namespace {
+
+template <typename T, std::enable_if_t<std::numeric_limits<T>::is_signed, void>* = nullptr>
+void assert_non_negative(T const& value)
+{
+  CUDF_EXPECTS(value >= T{0}, "Range scalar must be >= 0.");
+}
+
+template <typename T, std::enable_if_t<!std::numeric_limits<T>::is_signed, void>* = nullptr>
+void assert_non_negative(T const& value)
+{
+  // Unsigned values are non-negative.
```

Comment:
Is this done because we would get a compiler warning if we checked the sing of an unsigned variable?

Reply:
Yes, sir. That's the case.

Reply:
You should be able to replace the SFINAE with constexpr if, implement the assert funciton this way:
```
    if constexpr (std::is_signed<T>::value)
        CUDF_EXPECTS(value >= T{0}, "Range scalar must be >= 0.");
```

Reply:
Yay, C++17! Let me try that.

Reply:
Thanks for this suggestion, @vuule. (The switch to C++17 has been brilliant thus far.)